src/llmcompressor/modifiers/quantization/calibration.py

```diff
@@ -0,0 +1,251 @@
+    quantization_scheme = getattr(module, "quantization_scheme", None)
+    if not quantization_scheme:
+        # no quantization scheme nothing to do
+        return
+
+    # observers have a different lifecycle for kv_cache
+    if is_attention_module(module):
+        return
+
+    quantization_args = getattr(quantization_scheme, arg_name, None)
+    # dont need observers for dynamic
+    if quantization_args and not quantization_args.dynamic:
+        observer = quantization_args.get_observer()
+        observer = Observer.load_from_registry(
+            observer, quantization_args=quantization_args
+        )
+        module.register_module(f"{base_name}_observer", observer)
+
+
+def call_observer(module: Module, base_name: str, value: torch.Tensor):
+    """
+    Call a module's attached input/output observer using a provided value.
+    Update the module's scale and zp using the observer's return
+    values.
+
+    :param module: torch.nn.Module
+    :param base_name: substring used to fetch the observer, scales, and zp
+    :param value: torch.Tensor to be passed to the observer
+    """
+    observer = getattr(module, f"{base_name}_observer")
+    g_idx = getattr(module, "weight_g_idx", None)
+
+    updated_scale, updated_zero_point = observer(value, g_idx=g_idx)
+
+    # update scale and zero point
+    update_parameter_data(module, updated_scale, f"{base_name}_scale")
+    update_parameter_data(module, updated_zero_point, f"{base_name}_zero_point")
+
+
+def update_weight_zp_scale(module: Module):
+    """
+    marks a layer as ready for calibration which activates observers
+    to update scales and zero points on each forward pass
+
+    apply to full model with `model.apply(set_module_for_calibration)`
+
+    :param module: module to set for calibration
+    :param quantize_weights_upfront: whether to automatically
+       run weight quantization at the start of calibration
+    """
+    if not getattr(module, "quantization_scheme", None):
+        # no quantization scheme nothing to do
+        return
+
+    status = getattr(module, "quantization_status", None)
+    if not status or status != QuantizationStatus.INITIALIZED:
+        _LOGGER.warning(
+            f"Attempting set module with status {status} to calibration mode. "
+            f"but status is not {QuantizationStatus.INITIALIZED} - you may "
+            "be calibrating an uninitialized module which may fail or attempting "
+            "to re-calibrate a frozen module"
+        )
+
+    if module.quantization_scheme.weights is not None:
+        # set weight scale and zero_point up front, calibration data doesn't affect it
+        offloaded = is_module_offloaded(module)
+        if offloaded:
+            module._hf_hook.pre_forward(module)
+
+        call_observer(module=module, base_name="weight", value=module.weight)
+
+        if offloaded:
+            module._hf_hook.post_forward(module, None)
```

Comment:
not how often we use the hook wrt the hook, but another way is to use context manager

```python3
from contextlib import contextmanager

@contextmanager
def module_hook_context(module):
    offloaded = is_module_offloaded(module)
    if offloaded:
        module._hf_hook.pre_forward(module)
    try:
        yield
    finally:
        if offloaded:
            module._hf_hook.post_forward(module, None)

# Now the code can be simplified to:
with module_hook_context(module):
    call_observer(module=module, base_name="weight", value=module.weight)
```

Reply:
overall hook management is something we're targeting to clean-up later as we're introducing it in multiple places.